src/compiler/llvm/mir.rs

```diff
@@ -437,18 +437,48 @@ impl<'p, 'module, 'ctx>
         }
     }
 
+    fn declare_struct(&mut self, id: TypeId, path: &Path) -> Result<(), TransformerError> {
+        let label = self.to_label(path);
+        let struct_decl = self.context.opaque_struct_type(&label);
+        match self.ty_table.insert(id, struct_decl.into()) {
+            Some(_) => Err(TransformerError::TypeAlreadyDefined),
+            None => Ok(()),
+        }
+    }
+
     fn add_type(&mut self, id: TypeId, ty: &MirTypeDef) -> Result<(), TransformerError> {
         debug!("Adding a type to the Module");
 
-        let previous_value = ty
-            .into_basic_type_enum(self)
-            .and_then(|llvm_ty| self.ty_table.insert(id, llvm_ty));
-
-        // If `insert` returns `None` then it means there was no previous value associated with `id`
-        // otherwise, `id` was already defined and this should thrown an error.
-        match previous_value {
-            None => Ok(()),
-            Some(_) => Err(TransformerError::TypeAlreadyDefined),
+        // If type is already in the table then skip
+        if self.ty_table.contains_key(&id) {
+            // If this is a structure, then turn the declaration into a definition
+            match ty {
+                MirTypeDef::Structure { def, .. } => match def {
+                    MirStructDef::Defined(fields) => {
+                        // Get the structure declaration
+                        let s = self.ty_table.get_mut(&id).unwrap().into_struct_type();
+                        // Add fields to structure definition
+                        let field_types: Vec<_> = fields
+                            .iter()
+                            .map(|f| self.get_type(f.ty).unwrap().into_basic_type().unwrap())
+                            .collect();
+                        s.set_body(&field_types, false);
```

Comment:
Check to make sure that the structure does not already have a body defined, if it does, then return an error.  Otherwise, set the body and return `Ok`

Reply:
Sadly, there does not seem to be a way to check this.

Reply:
I can check to make sure that this has no fields, but that will miss the case where a struct is defined with no fields and then is defined again.